// new-find-package-json/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findSync = exports.find = exports.findAsync = void 0;
const tslib_1 = require("tslib");
const path = require("path");
const fs_1 = require("fs");
const debug = require("debug");
const log = debug('new-find-package-json:main');
/**
 * Find "package.json" files, starting from "input"
 * @param input Input Path to start searching from
 * @param base Base path, combined with input to form an absolute path (default: `process.cwd()`)
 * @param fileName The filename to search for (default: `package.json`)
 */
function findAsync(input, base, fileName) {
    return tslib_1.__asyncGenerator(this, arguments, function* findAsync_1() {
        log(`findSync: called with "${input}" and "${base}" and "${fileName}"`);
        base = !!base ? base : '';
        const useFileName = !!fileName ? fileName : 'package.json';
        const useBase = !!base || path.isAbsolute(base) ? base : process.cwd();
        let testPath = path.resolve(useBase, input);
        let wasRoot = false;
        while (testPath) {
            // Exectue search for "package.json" even on "/" | "C:"
            if (testPath === path.parse(testPath).root) {
                wasRoot = true;
            }
            const testFile = path.resolve(testPath, useFileName);
            log(`findSync: testing path "${testFile}"`);
            const result = yield tslib_1.__await(statPathAsync(testFile));
            if (!!result && result.isFile()) {
                log(`findSync: path exists and is file "${testFile}"`);
                yield yield tslib_1.__await(testFile);
                log(`findSync: after yield`);
            }
            // stop looping after having searched root, because there is no more going up
            if (wasRoot) {
                break;
            }
            testPath = path.resolve(testPath, '..');
        }
    });
}
exports.findAsync = findAsync;
exports.find = findAsync;
/**
 * Find "package.json" files, starting from "input"
 * @param input Input Path to start searching from
 * @param base Base path, combined with input to form an absolute path (default: `process.cwd()`)
 * @param fileName The filename to search for (default: `package.json`)
 */
function* findSync(input, base, fileName) {
    log(`findSync: called with "${input}" and "${base}" and "${fileName}"`);
    base = !!base ? base : '';
    const useFileName = !!fileName ? fileName : 'package.json';
    const useBase = !!base || path.isAbsolute(base) ? base : process.cwd();
    let testPath = path.resolve(useBase, input);
    let wasRoot = false;
    while (testPath) {
        // Exectue search for "package.json" even on "/" | "C:"
        if (testPath === path.parse(testPath).root) {
            wasRoot = true;
        }
        const testFile = path.resolve(testPath, useFileName);
        log(`findSync: testing path "${testFile}"`);
        const result = statPathSync(testFile);
        if (!!result && result.isFile()) {
            log(`findSync: path exists and is file "${testFile}"`);
            yield testFile;
            log(`findSync: after yield`);
        }
        // stop looping after having searched root, because there is no more going up
        if (wasRoot) {
            break;
        }
        testPath = path.resolve(testPath, '..');
    }
}
exports.findSync = findSync;
/**
 * Run "fs.promises.stat", but return "undefined" if error is "ENOENT"
 * follows symlinks
 * @param path The Path to Stat
 * @throws if the error is not "ENOENT"
 */
function statPathSync(path) {
    try {
        return fs_1.statSync(path);
    }
    catch (err) {
        if (err.code === 'ENOENT') {
            return undefined; // catch the error if the directory dosnt exist, without throwing an error
        }
        throw err;
    }
}
/**
 * Run "fs.promises.stat", but return "undefined" if error is "ENOENT"
 * follows symlinks
 * @param path The Path to Stat
 * @throws if the error is not "ENOENT"
 */
function statPathAsync(path) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return fs_1.promises.stat(path).catch((err) => {
            if (err.code === 'ENOENT') {
                return undefined; // catch the error if the directory dosnt exist, without throwing an error
            }
            throw err;
        });
    });
}
